doc/developer/design/20230330_recursion_limit.md

```diff
@@ -0,0 +1,178 @@
+
+## Extending SQL
+
+Here, I will discuss some more advanced approaches to express the limits instead of global variables. These are harder to implement, but have better usability in some complex situations for use case 2. (debugging), and would also support use case 3.
+
+There are several possibilities on how to extend SQL to allow for expressing the limits:
+- A. (OPTIONS clause on a SELECT),
+- B. QUERYOPTIONS clause on an entire query (this would be a new keyword to avoid confusion with the already existing OPTIONS clause, which is scoped to a SELECT),
+- C. on a WMR clause, e.g., `WITH MUTUALLY RECURSIVE MAX ITERATIONS 100`,
+- D. in a WHERE clause with a special, unmaterializable function, e.g., `WHERE iteration_number() < 100`.
+
+First of all, note that A. is not good:
+- If we simply want to set a limit for an entire query, i.e., set the same limit for all WMR clauses inside a single query, then B. is better than A., because a SELECT might appear as a subquery.
+- If we want a separate limit for each WMR clause, then it's better to put that limit right on the WMR clause (C.).
+
+The B., C., and D. possibilities are all somewhat suitable for all three use cases, because we can use them to set different limits in different queries, and also they make the limit part of the query's SQL definition, and therefore they survive restarts. However, they differ in usability for queries that have multiple WMR clauses, and also in implementation difficulties. I will discuss these differences in the two "Queries with multiple WMR clauses" sections ([usability](#Queries-with-multiple-WMR-clauses----usability) and [implementation difficulties](#queries-with-multiple-wmr-clauses----implementation-difficulties)).
+
+### SQL vs. global variables
+
+Before comparing the above SQL extension possibilities with each other, I first compare them with the global variables approach. I do this comparison by focusing on B., which is the least powerful of the SQL possibilities.
+
+For use case 1., B. has the usability advantage over a system variable that it avoids the danger that multiple users of the same system don't know about each other's limit needs, and one of them sets a lower limit than what the other would want. However, I guess users will typically just set an infinite limit whenever they touch the limit at all, so there won't be any significant conflicts. Also note that a usability advantage of the system variable approach is that it allows power users to opt out of the hard limit once and for all, i.e., avoid having to type it for every query. Taking into account both of these considerations, I would vote for a system variable.
+
+For use case 2., B. has the slight usability advantage over a session variable that it avoids the danger that someone forgets to reset the variable after a debugging session. However, I don't think this slight advantage would justify adding a new clause to the parser. **@MaterializeInc/surfaces**, would you agree?
```

Comment:
I'm personally not a huge fan of having a session variable affect the semantics of a query. I know there are other existing variables that can similarly affect the result of a query (`search_path`, `database`, `transaction_isolation`), but none of them fundamentally changes the behavior of how a query is executed.

It's also not entirely clear to me how the number of iterations is being set for views with WMR queries. Is it just the default value of the session variable? If so, that value can change in between versions, which means that a version upgrade can alter the behavior of an existing view which also feels wrong.

I would feel better if the number of iterations were explicitly in the query itself instead of derived from a session variable.

Reply:
> It's also not entirely clear to me how the number of iterations is being set for views with WMR queries. Is it just the default value of the session variable?

The session variable is for the debugging use case, which I meant mainly for one-off selects, during the development of a query.

> If so, that value can change in between versions, which means that a version upgrade can alter the behavior of an existing view which also feels wrong.

Hmm, yes, this sounds indeed a bit weird. Although the session variable's default value would be infinite in all versions, because that is for the debugging use case, but the system variable's default (which is for the "stopping divergent queries" use case) might indeed change between versions.

> I would feel better if the number of iterations were explicitly in the query itself instead of derived from a session variable.

I see your point!